src/Model/Resolver/Products/DataPostProcessor/Attributes.php

```diff
@@ -253,38 +254,43 @@ protected function appendWithOptions(
                 $values = array_flip( // Flip Array
                     array_merge( // phpcs:ignore
                         array_filter( // explode might return array with empty value, remove such values
-                            explode(',', $productAttributes[$id][$key]['attribute_value'] ?? '')
+                            explode(',', $productAttributes[$productId][$attributeCode]['attribute_value'] ?? '')
                         ),
                         $variantAttributeValues
                     )
                 );
 
-                $options = $attribute->getOptions();
-                array_shift($options);
-                $productAttributes[$id][$key]['attribute_options'] = [];
+                $productAttributes[$productId][$attributeCode]['attribute_options'] = [];
 
-                foreach ($options as $option) {
+                foreach ($attribute['attribute_options'] as $option) {
                     $value = $option->getValue();
 
                     if (!isset($values[$value])) {
                         continue;
                     }
 
                     $optionIds[] = $value;
-                    $productAttributes[$id][$key]['attribute_options'][$value] = [
+                    $productAttributes[$productId][$attributeCode]['attribute_options'][$value] = [
                         'value' => $value,
                         'label' => $option->getLabel()
                     ];
                 }
             }
         }
 
+        foreach ($attributes as $attributeCode => $attribute) {
+            // Collect all swatches (we will need additional data for them)
+            if ($this->swatchHelper->isSwatchAttribute($attribute)) {
```

Comment:
I remember this check was not very efficient. Not a problem unless very many attributes to load, but maybe see if something can be grouped there in M2 core. Not a priority though.